feat(blog): How DNS works in Firezone (#4828)
CloudFire-LLC/cloudfire: website/src/app/blog/how-dns-based-traffic-routing-works/readme.mdx

plague it, and how Firezone's DNS-based traffic routing was designed to address
them.

## A brief history lesson

But before we dive into all the fun details, it's helpful to first understand
the problem DNS was originally designed to solve.

DNS was created in the early 1980s to address a classic scalability problem with
the early ARPANET.

<Image
  src="/images/blog/how-dns-based-traffic-routing-works/arpanet-logical-map.png"
  alt="Early ARPANET logical map"
  width={1200}
  height={1200}
  className="mx-auto rounded shadow"
/>

<p className="text-sm italic text-center mx-auto">
  Figure 1: Early ARPANET logical map
</p>

The ARPANET, the precursor to the modern internet, was a network of computers
that connected various research institutions and government agencies across the
United States. It was the first network to use the new TCP/IP protocol suite,
which provided an IP address to each host on the network.

To connect to a host on the ARPANET, you needed its IP address. But there was a
problem: IP addresses are hard to remember, and it turns out humans aren't great
at remembering long strings of numbers. To solve this, a simple system was

> created to each host's IP address to a human-friendly "hostname".

Reviewer: ```suggestion
created to map each host's IP address to a human-friendly "hostname".
```

time) wanted to add a host to the file or update its IP address, it had to
contact the SRI (during standard business hours!) and file a formal request
containing the updated mapping.

Need to make an update over the weekend or on a US holiday? Too bad, you were
out of luck!

The system worked ok when there were only a handful of hosts to maintain. But,
as you might imagine, it became increasingly unmanageable as more hosts joined.

## Enter the Domain Name System (DNS)

Over the course of several meetings, early ARPANET engineers (who would later go
on to form the [IETF](https://www.ietf.org/)) devised a clever solution to the
scalability problem they were facing: instead of storing the hostname lookup
table in a single file, they would chop it up instead and distribute the entries
across several "Nameservers", each one responsible for maintaining the hosts
that belonged to a particular network, or "Domain".

Hence the name: the Domain Name System, or DNS for short.

### How DNS works

At a high level, DNS is a heirarchical system that distributes the
responsibility of resolving a a fully-qualified domain name (FQDN) to a series

> of nameservers, each one responsible for a resolving different part.

Reviewer: ```suggestion
of nameservers, each one responsible for resolving a different part.
```

<p className="text-sm italic text-center mx-auto">Figure 3: How DNS works</p>

> Here's how it works:

Reviewer: ```suggestion
Here's how it works nowadays:
```

1. If _this_ query misses the cache, the upstream resolver begins the full
   process of resolution. It starts by forwarding the query to a **root
   nameserver**. The root nameserver is at the top of the DNS hierarchy.
1. The root nameserver responds to the upsteam resolver with the IP address of
   the **TLD nameserver** for the root in question, for example `com` or `net`.
   The upstream resolver then forwards the query to the TLD nameserver.
1. The TLD nameserver responds with the IP address of the **authoritative
   nameserver** for the domain in question. This is the nameserver the owner of
   the domain will configure to respond to queries for that domain. It's where
   you typically configure your DNS records for your domain, for example.
1. The upstream resolver then forwards the query to the authoritative
   nameserver.
1. Finally, the authoritative nameserver responds with the IP address of the
   host in question, and the upstream resolver returns the final answer to the
   stub resolver on the host that originally made the query.
1. The application on the host can now connect to the IP address returned by the
   stub resolver.

> The system of root, TLD, and authoritate nameservers replace the function of the

Reviewer: ```suggestion
The system of root, TLD, and authoritative nameservers replace the function of the
```

original `HOSTS.TXT` file maintained by NIC in the ARPANET. However, now it's
distributed -- no one server or organization is responsible for maintaining the
entire database, and the system can scale as more hosts, domains, and IP
addresses are added.

On today's internet, the whole process for resolving an query usually takes only
a few hundred milliseconds. Caching resolvers help to speed up the process by
storing the results of queries for a certain amount of time, known as the
record's time-to-live (TTL).

This means that if a host makes the same query multiple times, the stub or
upstream resolver can return the result immediately (assuming the TTL hasn't
expired) without having to query the hierarchy of root, TLD, and authoritative

> nameservers again. This can speed up query times orders of magnitude, to the

Reviewer: ```suggestion
nameservers again. This can speed up query times by orders of magnitude, to the
```

taken by a malicious actor looking for potential entrypoints into an
organization.

Because of this issue, organizations often resort to running their own upstream
resolvers, configured to return different results (or sometimes none at all)
depending on the _source_ of the request. This technique is known as **Split
DNS**.

So how does it work?

## How Split DNS works

Split DNS is a technique wherein an orgnization maintains two (or more) separate
DNS servers (or a single server configured with two "zones") -- one for internal
resources and one for external resources. The internal server is only accessible
to users who are connected to an organization's internal network (such as a VPN
or branch office network), while the external zone is accessible to the outside
world.

<p className="text-sm italic text-center mx-auto">
  Figure 5: Split DNS in action
</p>

As an example, let's say an organization has an internal service called
`gitlab.company.com` which resolves to `10.10.10.10`. The organization's
internal DNS server would be configured to respond to queries for this service,
but the external DNS server would simply return `NXDOMAIN`, DNS speak for

> "Sorry, wrong number". This allows the organization to publish some records

Reviewer: ```suggestion
"Sorry, the number you have reached is not in service." This allows the organization to publish some records
```

"wrong number" makes me think someone picked up the phone, just not the person I expected. NXDOMAIN (AIUI) is more like, the call doesn't even reach any phone on the other end.

and you're set.

Split DNS works well for organizations that want to secure access to their own
internal applications, and continues to be a popular way to mitigate enumeration
attacks today.

So far, so good right?

### The problem with Split DNS

Split DNS works great when you have a clear distinction between external and
internal resources. It allows you to publish public addresses for your public
resources so anyone can access them, and publish private addresses for your
workforce so they can connect to the private resources that you manage.

Increasingly, however, cloud-delivered services are replacing their on-premise
equivalents across many organizations. The upside here is generally lower
operational cost -- pay the businesses making the software to host it for you as
opposed to hosting and managing it yourself, and reap the efficiency benefits in
the form of cost reduction.

But the downside with these public SaaS apps is that you can't use Split DNS to
hide them anymore -- they aren't running on your infrastructure, so there simply
aren't any internal zones and therefore no internal IPs to give to your

> workforce to connect to.

Reviewer 2: It's not clear here why would you want to hide a public service

Author: Fixed

As it turns out, there's a simple solution to this problem that's becoming more
and more common across popular SaaS providers. Many public SaaS apps like
GitHub, Slack, and Hubspot provide an IP allowlist feature that limits access to

> your tenant based on a range of allowed source addresses.

Reviewer 2: This doesn't exactly follow the previous paragraph, what this does is give you obvservability and control for traffic into your logical organization in the SaaS but it doesn't hide it.

Author: Yeah I think I should build this up a bit more -- how the shift to cloud exposes public SaaS apps, and how the IP allowlists provide some level of protection you once had by self-hosting the stuff.

Now, the studious reader will be quick to recognize the solution to our Split
DNS problem: can't the organization simply tunnel the resolved IP addresses of
these services through the VPN, just like they did before with internal

> services?

Reviewer 2: > tunnel the resolved IP addresses of these services through the VPN

sounds like you want to tunnel the DNS responses through the VPN but I think this paragraph refers to funneling traffic meant for that public service through a jump host inside your VPN.

Author: Yeah, I think I need to restructure this part. Split DNS's usefulness goes away when you access public SaaS apps.

But with the IP allowlist feature becoming more common, can Split DNS be resurrected somehow to help us lock down access like we had when self-hosting internally?

And yes, that can work in some circumstances, but there's one minor problem

> standing in our way: **virtual hosting**.

Reviewer 2: I think a bit more introduction to this paragraph is needed, something along the lines of:

"But there's a small problem if you simply route all traffic for the service's IP through your VPN jump host..."

If the service uses
[virtual hosting](https://en.wikipedia.org/wiki/Virtual_hosting), the IP address
that the DNS query resolves to is shared among many different services, and not
just the one the organization is interested in routing traffic for.

<p className="text-sm text-center mx-auto italic">
  Figure 5: Collateral damage can occur with Split DNS and some services
</p>

> This means that an organization wishing to route DNS traffc _only_ for

Reviewer 2: I'd add something like:

If we were to just add github's IP to the OS's routing table all traffic for that IP will be going through the VPN regardless whether the traffic is meant for github or any other service sitting behind the same IP

Author: I added a simplified version of this

`github.com` might inadvertently route traffic for _other_ GitHub (and maybe
even non-GitHub) services as well!

#### NAT to the rescue?

Ok, but what about another solution: can the organization simply configure its
client devices to use the organization's DNS server to respond with internal IP
addresses for the public services?

That's much better -- it solves the problem of routing traffic for the wrong
service, but it introduces a new problem: the organization now needs to maintain
a list of all the public resources its workforce needs to access, then configure
its DNS server to respond with a unique internal IP address for each one,

> finally it needs to be reasonably sure these IP addresses don't conflict with

Reviewer: ```suggestion
and finally it needs to be reasonably sure these IP addresses don't conflict with
```

other networks available to the employee's machine, such as a home printer.

<p className="text-sm text-center mx-auto italic">
  Figure 6: Split DNS with NAT: A management nightmare
</p>

On top of the addressing headache, the organization needs to operate NAT routers
that accept the traffic from the VPN clients across their workforce, translate
the source addresses so they appear to come from one of the IPs in the
allowlist, and finally forward the traffic on to the service in question. The IP

> addressing configuration and NAT router configuration would need to be kept in

Reviewer 2: I'm not clear on what "The IP addressing configuration" means here

Reviewer 2: Furthermore, the problem that an standard nat gateway doesn't solve is translating the destination address!

Author: Yeah I'll probably avoid going into the details of NAT here, and just mention how it's pain to manage all this stuff manually.

sync so that they function together.

So we've taken what was originally a DNS problem and translated it to a
configuration management problem. So let's solve it!

## DNS-based traffic routing

By now you're probably thinking, "this is all well and good, but what does this
have to do with Firezone?". Glad you asked!

Firezone's approach to DNS was designed to address both the insecurity of DNS

> and the shortcomings of Split DNS, into a single solution that's just as secure

Reviewer: ```suggestion
and the shortcomings of Split DNS, resulting in a single solution that's just as secure
```

as Split DNS, but much more flexible and easier to manage.

### How it works

When a user signs in, the Firezone Client configures its host operating system
to use a special, lightweight DNS proxy running inside the Client for all DNS
queries on the system.

This proxy lives at `100.100.111.0/24` for IPv4 queries and
`fd00:2021:1111:8000:100:100:111:0/120` for IPv6 queries. These addresses are
not routable on the public internet, so there's no risk that queries for
internal resources will leave the device unexpectedly.

The IP addresses the proxy chooses to "listen on" are determined by the
[upstream resolvers](#configuring-client-dns-upstream-resolvers) available to
the Client. For each IPv4 upstream resolver, the proxy listens on an address in
the `100.100.111.0/24` range, and for each IPv6 upstream resolver, the proxy
listens on an address in the `fd00:2021:1111:8000:100:100:111:0/120` range.

<p className="text-sm text-center mx-auto italic">
  Figure 7: Firezone's embedded stub resolver
</p>

> When an application makes a DNS query, the proxy checks if the name matches a

Reviewer 2: Saying that the proxy checks it is kind of not precise, connlib does the checking, we can say in the first paragraphthat "the client" intercept DNS requests and then forwards the queries to an embedded resolver(hickory!)

Author: I mention it's a stub resolver with "an interesting property" or something along those lines

Firezone Resource that the user has access to. If it does, the Client then

> requests a Gateway serving that Resource to resolve the DNS query. When the

Reviewer 2: Perhaps explain first if it doesn't? Since that case is quite simpler

Reviewer 2: Might be worth saying it is requesting both access and DNS resolution at the same time

Author: Good point, I've got that now. I'm still working on how to detail how it works in a digestible way.

> Gateway responds, the proxy will create a temporary mapping between the actual

Reviewer 2: The reader might have no context as to what a Firezone gateway is, we might need an small section with Firezone's context

Author: Good point, added that in an `<Alert>` callout

> IP address of the Resource and the dummy IP address it returned to the

Reviewer 2: The "proxy"(I'm guessing hickory) is completely oblivious to the dummy ips, this is the job of connlib

Reviewer 2: might be worth mentioning that "the dummy IP" also depends the gateway peer, meaning that we can distinguish between different services with same ip on different gateways.

Author: Yeah I think I'll use the term "stub resolver" instead of proxy here... and just say "Client includes a zero-trust aware stub resolver" or something? Should we make a better marketing name for it, or just call it connlib (and I introduce connlib here too?)

I was hoping to avoid exposing some of the details/naming that don't matter, but some of this stuff could be interesting

application. From this point onward, all subsequent packets to the dummy IP
address are forwarded to the Gateway that resolved the query. These IPs are

> persisted for the duration of the Client's session.

Reviewer 2: also we might mention the ips are mangled on the client?

Author: Yeah will add that detail today

```
user@host:~ % dig ifconfig.net

; <<>> DiG 9.10.6 <<>> ifconfig.net
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 51756
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;ifconfig.net.			IN	A

;; ANSWER SECTION:
ifconfig.net.		300	IN	A	100.96.0.2
ifconfig.net.		300	IN	A	100.96.0.1

;; Query time: 262 msec
;; SERVER: fd00:2021:1111:8000:100:100:111:0#53(fd00:2021:1111:8000:100:100:111:0)
;; WHEN: Sat Feb 17 14:04:40 PST 2024
;; MSG SIZE  rcvd: 86
```

<p className="text-sm text-center mx-auto italic">
  Figure 8: An example dig command with a DNS-based Resource
</p>

If the proxy sees a query for a name that **is not** a Firezone Resource it will
forward the query to the host's default stub resolver, as if Firezone never

> existed. This means that Clients are automatically configured for Split DNS in

Reviewer 2: Not really, the host's stub resolver is not used, it's the DNS's server that is used.

Reviewer 2: Also it might be interesting telling how we resolved the timeouts and multiple DNS servers

Author: Yeah those will be good to mention too, probably with a graphical aid

Firezone -- no other configuration is necessary other than adding the desired
Resources in the admin portal.

> {/* TODO: Add much more detail here */}

Author: Planning to add much more in this section about how our implementation works.
